### PR TITLE
slave: monitor: add Associated STA Link Metrics Response

### PR DESCRIPTION
**Pull request by vitalii-komisarenko**
At the moment Associated STA Link Metrics are handled as
a part of *_HOSTAP_STATS_MEASUREMENT_* messages.
We would like to handle them separately.
    
This commit adds a flow for the Associated STA Link Metrics Response
that basically duplicates a subset of *_HOSTAP_STATS_MEASUREMENT_*.
It is a preparation for removing above-mentioned subset from Hostap Measurements.

### CODE REVIEW

**Comment by vitalii-komisarenko** [comment]
Replaced by  PR #1149

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 1]
```diff
@@ -179,11 +181,50 @@ void monitor_stats::process()
             sta_stats_msg.rx_load_percent   = sta_stats.rx_load_percent_curr;
             sta_stats_msg.stats_delta_ms    = sta_stats.delta_ms;
             sta_stats_msg.rx_rssi           = sta_stats.rx_rssi_curr;
+
+            beerocks_message::sBssidInfo bss_info;
+            bss_info.bssid                      = network_utils::mac_from_string(sta_mac);
```
Comment (vitalybu) [inline agent/src/beerocks/monitor/monitor_stats.cpp]: `bssid` should be the MAC address of the AP not the STA.
Unless the field name is incorrect

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 2]
```diff
@@ -2088,6 +2095,56 @@ bool backhaul_manager::get_sta_bssid(const sMacAddr &sta_mac, sMacAddr &bssid)
     return false;
 }
 
+bool backhaul_manager::handle_associated_sta_link_metrics_query(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                                const std::string &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received ASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE , mid=" << std::dec
```
Comment (vitalybu) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2130]:
```suggestion
    LOG(DEBUG) << "Received ASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE, mid=" << std::dec
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 3]
```diff
@@ -2088,6 +2095,56 @@ bool backhaul_manager::get_sta_bssid(const sMacAddr &sta_mac, sMacAddr &bssid)
...
+        error_code_tlv->sta_mac() = mac->sta_mac();
+
+        LOG(DEBUG) << "Send a ASSOCIATED_STA_LINK_METRICS_RESPONSE_MESSAGE back to controller";
+        return send_cmdu_to_bus(cmdu_tx, src_mac, bridge_info.mac);
+    } else {
+        LOG(DEBUG) << "client with mac address " << mac->sta_mac() << " connected to " << bssid;
+        //TODO
+        return false;
+    }
+}
```
Comment (vitalybu) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]:
```suggestion
    }
    
    LOG(DEBUG) << "client with mac address " << mac->sta_mac() << " connected to " << bssid;
    //TODO
    return false;
}
```
`//TODO` what?

Also, if you already found the `bssid` to which the STA is connected, it's better to forward the CMDU the correct slave from here. Otherwise (by returning `false`) the message will be forwarded to all the slaves.

File: tests/test_flows.py [hunk 4]
```diff
@@ -488,6 +488,18 @@ def test_client_association_dummy(self):
         self.check_log(env.agents[0].radios[0],
                        r"Got client disallow request for {}".format(sta.mac))
 
+    def test_client_association_link_metrics(self):
```
Comment (vitalybu) [inline tests/test_flows.py:548]: Please this change to a separate commit

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 5]
```diff
@@ -2080,13 +2080,17 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     }
 }
 
-bool backhaul_manager::get_sta_bssid(const sMacAddr &sta_mac, sMacAddr &bssid)
+bool backhaul_manager::get_sta_bssid_and_socket(const sMacAddr &sta_mac, sMacAddr &bssid,
+                                                Socket **socket)
```
Comment (vitalybu) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: It better to use a reference to a pointer (https://stackoverflow.com/a/3834103)
```suggestion
                                                Socket *&socket)
```

But why do you need to return both the `bssid` and the slave socket?
You can return the `sRadioInfo` to which the STA is connected and then extract both the `bssid` and the slave socket from there.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 6]
```diff
@@ -2080,13 +2080,17 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
...
 {
     for (const auto &slave : slaves_sockets) {
         auto associated_clients_map = slave->associated_clients_map;
         for (const auto &vap : associated_clients_map) {
             if (vap.second.find(sta_mac) != vap.second.end()) {
                 bssid = vap.first;
+                if (socket) {
+                    *socket = slave->slave;
+                }
```
Comment (vitalybu) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Assuming a reference to a pointer:
```suggestion
                socket = slave->slave;
```

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 7]
```diff
@@ -1411,6 +1411,29 @@ bool slave_thread::handle_cmdu_backhaul_manager_message(
 
         break;
     }
+    case beerocks_message::ACTION_BACKHAUL_HOSTAP_STATS_MEASUREMENT_REQUEST: {
+        LOG(DEBUG) << "ACTION_BACKHAUL_HOSTAP_STATS_MEASUREMENT_REQUEST";
+        if (monitor_socket) {
```
Comment (vitalybu) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: According to our coding standards:
```suggestion
        if (!monitor_socket) {
            break;
        }
        ...        
```

File: tests/test_flows.py [hunk 8]
```diff
@@ -499,6 +499,19 @@ def test_client_association_link_metrics(self):
                                      tlv(0x95, 0x0006, '{sta_mac}'.format(sta_mac=sta_mac)))
         self.check_log(env.agents[0],
                        "client with mac address {sta_mac} not found".format(sta_mac=sta_mac))
+        time.sleep(1)
```
Comment (vitalybu) [inline tests/test_flows.py:559]: Again, tests should be updated in a separate commit

**Review by vitalybu — DISMISSED** [review]

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 9]
```diff
@@ -2073,6 +2073,25 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     }
 }
 
+bool backhaul_manager::get_sta_bssid(
+    const std::unordered_map<sMacAddr, associated_clients_t> &clients_map, const sMacAddr &sta_mac,
+    sMacAddr &bssid)
+{
+    return std::any_of(clients_map.begin(), clients_map.end(),
```
Comment (morantr) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Can't it be done with ranged for-loops? They are more readable and have less indentation.
It is mention as preferred on https://github.com/prplfoundation/prplMesh/blob/master/CODINGSTYLE.md#loops.

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 10]
```diff
@@ -1411,6 +1411,29 @@ bool slave_thread::handle_cmdu_backhaul_manager_message(
...
+            auto request_in = beerocks_header->addClass<
+                beerocks_message::cACTION_BACKHAUL_HOSTAP_STATS_MEASUREMENT_REQUEST>();
+            if (request_in == nullptr) {
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: With a smart pointer, prefer doing `if (!smart_ptr)` over `if (smart_ptr == nullptr)`

File: agent/src/beerocks/slave/son_slave_thread.cpp [hunk 11]
```diff
@@ -1411,6 +1411,29 @@ bool slave_thread::handle_cmdu_backhaul_manager_message(
...
+
+            auto request_out = message_com::create_vs_message<
+                beerocks_message::cACTION_MONITOR_HOSTAP_STATS_MEASUREMENT_REQUEST>(
+                cmdu_tx, beerocks_header->id());
+            if (request_out == nullptr) {
+                LOG(ERROR) << "Failed building message!";
+                return false;
+            }
+            request_out->sync() = request_in->sync();
+            LOG(DEBUG) << "Will now send ACTION_MONITOR_HOSTAP_STATS_MEASUREMENT_REQUEST";
```
Comment (morantr) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: the "will now" is redundant.
`LOG(DEBUG) << "send ACTION_MONITOR_HOSTAP_STATS_MEASUREMENT_REQUEST";`

**Review by morantr — APPROVED** [review]
Nothing major so approving.
I really don't like the use of std::any instead of ranged loops.
also, lasts commit "cppcheck: existing issues: add unfixable issues for associated sta li……nk metrics" headline is too long.

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 12]
```diff
@@ -179,11 +181,53 @@ void monitor_stats::process()
...
+            bss_info.earliest_measurement_delta = sta_stats.delta_ms;
+            // TODO: MAC data rate and Phy rate are not necessarily the same
```
Comment (tomereli) [inline agent/src/beerocks/monitor/monitor_stats.cpp]: TODO should be accompanied with a github issue URL

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 13]
```diff
@@ -179,11 +181,53 @@ void monitor_stats::process()
...
         beerocks_header->actionhdr()->id() = requests_list.front();
         requests_list.pop_front();
         message_com::send_cmdu(slave_socket, cmdu_tx);
+
+        LOG(DEBUG) << "Preparing ACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE. "
+                   << "Number of entries: " << bss_infos.size();
+
+        auto sta_metrics = message_com::create_vs_message<
+            beerocks_message::cACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE>(cmdu_tx);
+        if (sta_metrics == nullptr) {
```
Comment (tomereli) [inline agent/src/beerocks/monitor/monitor_stats.cpp]:
```suggestion
        if (!sta_metrics) {
```

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 14]
```diff
@@ -179,11 +181,53 @@ void monitor_stats::process()
...
+            LOG(ERROR) << "Failed building "
+                          "cACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE message!";
+            return;
+        }
+
+        if (!sta_metrics->alloc_bssid_info_list(bss_infos.size())) {
+            LOG(ERROR) << "Failed allocate_bssid_info_list";
```
Comment (tomereli) [inline agent/src/beerocks/monitor/monitor_stats.cpp]:
```suggestion
            LOG(ERROR) << "Failed allocate_bssid_info_list";
            return;
```

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 15]
```diff
@@ -179,11 +181,53 @@ void monitor_stats::process()
...
+        }
+
+        // TODO: set sta_metrics->sta_mac
```
Comment (tomereli) [inline agent/src/beerocks/monitor/monitor_stats.cpp]: TODO is both unclear and missing link to a github issue.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 9 above)
Comment (tomereli) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: You are right Moran, but `cppcheck` thinks otherwise :/

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 16]
```diff
@@ -2073,6 +2073,25 @@ bool backhaul_manager::handle_1905_1_message(ieee1905_1::CmduMessageRx &cmdu_rx,
     }
 }
 
+bool backhaul_manager::get_sta_bssid(
```
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Another option is to return sMacAddr instead of having it as a parameter - we can return an `network_utils::ZERO_MAC` if not found (nitpick).

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 17]
```diff
@@ -2091,11 +2110,10 @@ bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx
     bool associated_client_found = false;
     for (const auto &slave : slaves_sockets) {
         auto associated_clients_map = slave->associated_clients_map;
-        for (const auto &vap : associated_clients_map) {
-            if (vap.second.find(client_info_tlv_r->client_mac()) != vap.second.end()) {
-                associated_client_found = true;
-                break;
-            }
+        sMacAddr bssid;
```
Comment (tomereli) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: So this will change to:

```
auto bssid = get_sta_bssid(associated_clients_map, client_info_tlv_r->client_mac());
if (bssid != network_utils::ZERO_MAC) {
  associated_client_found = true;
  break;
}
```
(nitpick continued :) )

**Review by tomereli — COMMENTED** [review]
Not finished.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 16 above)
Comment (vitalii-komisarenko) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: shall `ZERO_MAC` convert implicitly to `false` in boolean context? :)

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 9 above)
Comment (arnout) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: >> Can't it be done with ranged for-loops? They are more readable and have less indentation.

That they are more readable is a matter of taste.

>> It is mention as preferred on https://github.com/prplfoundation/prplMesh/blob/master/CODINGSTYLE.md#loops.

That is only compared to "normal" for-loops, not compared to std::algorithms.


> You are right Moran, but `cppcheck` thinks otherwise :/

That's not good enough of a reason. If a tool doesn't do what we want, we have to change the tool, not what we want.

That said, I guess the main argument against using any_of is that it's too verbose? There are actually two ways that we can reduce the verbosity: by using C++14 auto, and by defining wrappers for the algorithms that take a container rather than an iterator.

```
    return any_of(clients_map, [&](const auto &vap) {
        return any_of(vap.second, [&](const auto &mac) {
            if (mac.first == sta_mac) {
                bssid = vap.first;
                return true;
            }
            return false;
        });
    });
```
where
```
template <class T, class UnaryPredicate>
bool any_of(const T& container, UnaryPredicate p)
{
    return any_of(std::begin(container), std::end(container), p);
}
```

Without that helper function, it's still not that bad if you use auto:

```
    return std::any_of(clients_map.begin(), clients_map.end(), [&](const auto &vap) {
        return std::any_of(vap.second.begin(), vap.second.end(), [&](const auto &mac) {
            if (mac.first == sta_mac) {
                bssid = vap.first;
                return true;
            }
            return false;
        });
    });
```

but I agree that it becomes bad when the lambda declaration no longer fits on the same line.

So I think the indentation argument is not really valid. Thus, the question is: which of the two is more readable?

I think we'll need to take a vote on #democracy on this.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml [hunk 18]
```diff
@@ -976,3 +976,16 @@ sChannelScanResults:
   channel_utilization:
     _comment: Indicates the fraction of the time AP senses that the channel is in use by the neighboring AP for transmissions.
     _type: uint32_t
+
+sBssidInfo:
```
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:974]: (nitpick) I don't think the name is very good. It's not info about a BSS, it's the STA link metrics for one particular BSS... Maybe `sSingleSTALinkMetric`?

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_monitor.yaml [hunk 19]
```diff
@@ -90,6 +90,18 @@ cACTION_MONITOR_HOSTAP_ACTIVITY_NOTIFICATION:
   _type: class
   params: sApActivityNotificationParams 
 
+cACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE:
```
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_monitor.yaml:98]: In #1198 we decided to use the IEEE1905.1 message type directly rather than defining a new TLV that copies everything from the EasyMesh message. You could do the same here...

That said, it's probably a lot of work to rework this big PR to do that now. So I'm not going to say you have to do it.

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 20]
```diff
@@ -220,11 +222,65 @@ void monitor_stats::process()
...
+        LOG(DEBUG) << "Preparing ACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE. "
+                   << "Number of entries: " << bss_infos.size();
+
+        std::for_each(
+            bss_infos.begin(), bss_infos.end(),
+            [&](std::pair<sMacAddr, std::vector<beerocks_message::sBssidInfo>> bss_info) {
+                auto sta_metrics = message_com::create_vs_message<
+                    beerocks_message::cACTION_MONITOR_CLIENT_ASSOCIATED_STA_LINK_METRIC_RESPONSE>(
+                    cmdu_tx, message_id);
+                if (sta_metrics == nullptr) {
```
Comment (arnout) [inline agent/src/beerocks/monitor/monitor_stats.cpp]:
```suggestion
                if (!sta_metrics) {
```
https://github.com/prplfoundation/prplMesh/blob/master/CODINGSTYLE.md#klocwork

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 21]
```diff
@@ -2146,15 +2156,7 @@ bool backhaul_manager::handle_client_capability_query(ieee1905_1::CmduMessageRx
...
-    bool associated_client_found = false;
-    for (const auto &slave : slaves_sockets) {
-        auto associated_clients_map = slave->associated_clients_map;
-        if (get_sta_bssid(associated_clients_map, client_info_tlv_r->client_mac()) !=
-                network_utils::ZERO_MAC) {
-            associated_client_found = true;
-            break;
-        }
-    }
+    bool associated_client_found = (bool)get_sta_radio(client_info_tlv_r->client_mac());
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: Please no C-style casts. I believe this works:
```suggestion
    bool associated_client_found { get_sta_radio(client_info_tlv_r->client_mac()) };
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 22]
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
     return radio != slaves_sockets.end() ? *radio : nullptr;
 }
 
+bool backhaul_manager::handle_associated_sta_link_metrics_query(ieee1905_1::CmduMessageRx &cmdu_rx,
+                                                                const std::string &src_mac)
+{
+    const auto mid = cmdu_rx.getMessageId();
+    LOG(DEBUG) << "Received ASSOCIATED_STA_LINK_METRICS_QUERY_MESSAGE , mid=" << std::dec
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2130]: I believe the std::dec shouldn't be needed any more because we reset the formatters in the LOG destructor, no?

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 23]
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
...
+               << int(mid);
+
+    if (!cmdu_tx.create(mid,
+                        ieee1905_1::eMessageType::ASSOCIATED_STA_LINK_METRICS_RESPONSE_MESSAGE)) {
+        LOG(ERROR)
+            << "cmdu creation of type ASSOCIATED_STA_LINK_METRICS_RESPONSE_MESSAGE, has failed";
+        return false;
+    }
+
+    auto mac = cmdu_rx.getClass<wfa_map::tlvStaMacAddressType>();
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2140]: (nitpick) mac can be anything, sta_mac would be better.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 24]
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
...
+    if (!mac) {
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2141]: (nitpick) do all checks before the cmdu_tx::create.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 25]
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
...
+    if (!assoc_link_metrics) {
+        LOG(ERROR) << "Failed to create tlvAssociatedStaLinkMetrics tlv";
+        return false;
+    }
+
+    // Check if it is an error scenario - if the STA specified in the STA link Query message is not associated
+    // with any of the BSS operated by the Multi-AP Agent
+    std::shared_ptr<sRadioInfo> radio = get_sta_radio(mac->sta_mac());
+    if (!radio) {
+        LOG(ERROR) << "client with mac address " << mac->sta_mac() << " not found";
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2156]: This is not really an error, rather info or warning. It is expected to happen, occasionally, because the controller may send the request just as the station is disassociating.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 26]
```diff
@@ -2141,6 +2148,61 @@ backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
...
+        return send_cmdu_to_bus(cmdu_tx, src_mac, bridge_info.mac);
+    } else {
+        sMacAddr bssid = get_sta_bssid(radio->associated_clients_map, mac->sta_mac());
+        if (bssid == network_utils::ZERO_MAC) {
+            LOG(ERROR) << "Cannot find sta bssid";
+            return false;
+        }
+        LOG(DEBUG) << "client with mac address " << mac->sta_mac() << " connected to " << bssid;
+        //TODO: non-error flow for Associated STA Link Metrics Request
+        return false;
```
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: (nitpick) since you already created the response message, I'd still send it here but without the required TLV. But since you fix that later in the same PR it's not important.

File: tests/test_flows.py [hunk 27]
```diff
@@ -531,6 +531,18 @@ def test_client_association_dummy(self):
         self.check_log(env.agents[0].radios[0],
                        r"Got client disallow request for {}".format(sta.mac))
 
+    def test_client_association_link_metrics(self):
+        ''' This test verifies that a MAUT with an associated STA responds to
+        an Associated STA Link Metrics Query message with an Associated STA Link Metrics
+        Response message containing an Associated STA Link Metrics TLV for the associated STA.'''
+
+        sta_mac = "11:11:33:44:55:66"
```
Comment (arnout) [inline tests/test_flows.py:553]:
```suggestion
        sta = env.Station.create()
```

File: tests/test_flows.py [hunk 28]
```diff
@@ -531,6 +531,18 @@ def test_client_association_dummy(self):
...
+        debug("Send link metrics query for unconnected STA")
+        env.controller.dev_send_1905(env.agents[0].mac, 0x800D,
+                                     tlv(0x95, 0x0006, '{sta_mac}'.format(sta_mac=sta_mac)))
```
Comment (arnout) [inline tests/test_flows.py:556]:
```suggestion
                                     tlv(0x95, 0x0006, '{}'.format(sta.mac)))
```
The `{var}` form is very useful in f-strings, but without them it's just too verbose IMO.

File: tests/test_flows.py [hunk 29]
```diff
@@ -531,6 +531,18 @@ def test_client_association_dummy(self):
...
+        self.check_log(env.agents[0],
```
Comment (arnout) [inline tests/test_flows.py:557]: We should actually check the error TLV in the reply, but that requires #1142 of course...

File: tests/test_flows.py [hunk 30]
```diff
@@ -542,6 +542,19 @@ def test_client_association_link_metrics(self):
                                      tlv(0x95, 0x0006, '{sta_mac}'.format(sta_mac=sta_mac)))
         self.check_log(env.agents[0],
                        "client with mac address {sta_mac} not found".format(sta_mac=sta_mac))
+        time.sleep(1)
+
+        sta = env.Station.create()
```
Comment (arnout) [inline tests/test_flows.py:561]: To distinguish from the previous sta, use sta1 and sta2.

File: ci/cppcheck/cppcheck_existing_issues.txt [hunk 31]
```diff
@@ -678,3 +678,7 @@
 '/builds/prpl-foundation/prplMesh/controller/src/beerocks/master/son_management.cpp: style: Variable 'op_error_code' is assigned a value that is never used. [unreadVariable]        auto op_error_code = eChannelScanOpErrCode::CHANNEL_SCAN_OP_SUCCESS;
 '/builds/prpl-foundation/prplMesh/controller/src/beerocks/master/son_management.cpp: style: Variable 'op_error_code' is assigned a value that is never used. [unreadVariable]            op_error_code = eChannelScanOpErrCode::CHANNEL_SCAN_OP_SCAN_IN_PROGRESS;
 '/builds/prpl-foundation/prplMesh/controller/src/beerocks/master/son_management.cpp: style: Variable 'result_status' is assigned a value that is never used. [unreadVariable]            result_status = last_scan_success;
+'/builds/prpl-foundation/prplMesh/agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h: warning: Virtual function 'on_thread_stop' is called from destructor '~backhaul_manager()' at line 332. Dynamic binding is not used. [virtualCallInConstructor]    virtual void on_thread_stop() override;
```
Comment (arnout) [inline ci/cppcheck/cppcheck_existing_issues.txt]: This one was already in there, it shouldn't be duplicated.

To regenerate this file, just run ci/cppcheck/cppcheck-new-issues.sh manually and copy cppcheck_results.txt to this file. Or better yet let CI do the job and take it from the cppcheck-diff artefacts.

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 32]
```diff
@@ -220,11 +222,65 @@ void monitor_stats::process()
             sta_stats_msg.rx_rssi           = sta_stats.rx_rssi_curr;
 
             sta_count++;
+
+            beerocks_message::sBssidInfo bss_info;
```
Comment (arnout) [inline agent/src/beerocks/monitor/monitor_stats.cpp]: You should also get bssid here. It's a bit hard to get, because the sta_node only has a vap_id:
```
    bss_info.bssid = network_utils::mac_from_string(mon_db->vap_get_by_id(sta_node->get_vap_id())->get_mac());
```
and you need to handle the possible nullpointer from vap_get_by_id.

But setting bssid this way removes the need for the backhaul manager to be involved in the response path.

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 33]
```diff
@@ -51,7 +51,7 @@ void monitor_stats::add_request(uint16_t id, uint8_t sync)
         }
         next_poll_id = mon_db->get_poll_id() + 1;
     }
-    requests_list.push_back(id);
+    requests_list.push_back(sMeasurementsRequest(id, sta_mac));
```
Comment (arnout) [inline agent/src/beerocks/monitor/monitor_stats.cpp:54]:
```suggestion
    requests_list.push_back({ id, sta_mac });
```
With that, the constructor is not needed.

File: agent/src/beerocks/monitor/monitor_stats.cpp [hunk 34]
```diff
@@ -54,6 +54,159 @@ void monitor_stats::add_request(uint16_t id, uint8_t sync, const sMacAddr &sta_m
...
+    beerocks_header->actionhdr()->id() = request.message_id;
+    message_com::send_cmdu(slave_socket, cmdu_tx);
+}
+
+void monitor_stats::send_associated_sta_link_metrics(const sMeasurementsRequest &request)
+{
+    for (auto it = mon_db->sta_begin(); it != mon_db->sta_end(); ++it) {
+        auto sta_mac  = it->first;
+        auto sta_node = it->second;
+        if (sta_node == nullptr) {
```
Comment (arnout) [inline agent/src/beerocks/monitor/monitor_stats.cpp]:
```suggestion
        if (!sta_node) {
```

**Review by arnout — APPROVED** [review]
Although I have a lot of nitpick comments, there is only one that is important enough to be absolutely required to change: don't use C-style cast.

Feel free to ignore all the other comments.

I propose to merge this PR even though it doesn't pass yet in certification.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp [hunk 35]
```diff
@@ -2142,6 +2142,16 @@ sMacAddr backhaul_manager::get_sta_bssid(
     return bssid;
 }
 
+std::shared_ptr<backhaul_manager::sRadioInfo>
+backhaul_manager::get_sta_radio(const sMacAddr &sta_mac)
+{
+    auto radio = std::find_if(
+        slaves_sockets.begin(), slaves_sockets.end(), [&sta_mac](std::shared_ptr<sRadioInfo> r) {
```
Comment (mariomaz) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2120]: I think that parameter to predicate function should be const reference
```
const std::shared_ptr<sRadioInfo> &radio_info
```

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 22 above)
Comment (mariomaz) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2130]: Most of the times the mid value is formatted as hex with `std::hex` while logging

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp (hunk 35 above)
Comment (rmelotte) [reply agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2120]: Good point. 
Also, since it does not share the ownership of the sRadioInfo, I don't see why it couldn't be a `sRadioInfo &` instead of a shared_ptr.

**Review by rmelotte — APPROVED** [review]
:+1: 
There are still some nitpick comments remaining, but as this PR is needed for https://github.com/prplfoundation/prplMesh/issues/394 and is a giant leap forward for 4.7.7, let's merge it!
